src/garage/sampler/base.py

```diff
@@ -1,43 +1,80 @@
+"""Base sampler class."""
+
 import abc
+import copy
 
 
 class Sampler(abc.ABC):
-    @abc.abstractmethod
-    def start_worker(self):
-        """Initialize the sampler,
+    """Base class of all samplers."""
 
-        e.g. launching parallel workers if necessary.
+    def __init__(self, algo, env):
+        """Construct a Sampler from an Algorithm.
+
+        Calling this method is deprecated.
         """
-        pass
+        self.algo = algo
+        self.env = env
 
-    @abc.abstractmethod
-    def obtain_samples(self, itr):
-        """Collect samples for the given iteration number.
+    @classmethod
+    def construct(cls, worker_factory, agents, envs):
+        """Construct this sampler.
 
         Args:
-            itr (int): Iteration number.
+            worker_factory(WorkerFactory): Pickleable factory for creating
+                workers. Should be transmitted to other processes / nodes where
+                work needs to be done, then workers should be constructed
+                there.
+            agents(Agent or List[Agent]): Agent(s) to use to perform rollouts.
+                If a list is passed in, it must have length exactly
+                `worker_factory.n_workers`, and will be spread across the
+                workers.
+            envs(gym.Env or List[gym.Env]): Environment rollouts are performed
+                in. If a list is passed in, it must have length exactly
+                `worker_factory.n_workers`, and will be spread across the
+                workers.
 
         Returns:
-            list[dict]: A list of paths.
+            Sampler: An instance of `cls`.
 
         """
-        pass
+        # This implementation works for most current implementations.
+        # Relying on this implementation is deprecated, but calling this method
+        # is not.
+        fake_algo = copy.copy(worker_factory)
+        fake_algo.policy = agents
+        return cls(fake_algo, envs)
 
-    @abc.abstractmethod
-    def shutdown_worker(self):
-        """Terminate workers if necessary."""
-        pass
+    def start_worker(self):
+        """Initialize the sampler.
 
+        i.e. launching parallel workers if necessary.
 
-class BaseSampler(Sampler):
-    """Base class for sampler.
+        This method is deprecated, please launch workers in construct instead.
```

Comment:
perhaps you can raise a warning here for some extra nagging?

Reply:
I don't think that's quite warranted yet, since it's going to nag all the time.